src/ndarray/ndarray.cc

```diff
@@ -168,16 +168,28 @@ nnvm::Symbol NDArray::get_autograd_symbol() const {
 
 #if MXNET_USE_MKLDNN == 1
 
-NDArray::NDArray(const mkldnn::memory *mkldnn_mem, bool static_data)
```

Comment:
it seems the old impl is incorrect. what if `static_data=false`?

Reply:
If static_data=false, then NDArray will take responsible to free this memory at deconstruction. But it has a double free risk that if mkldnn_mem be freed outside. I agreed that it's not a good design, so I changed it to std::shared_ptr<mkldnn::memory> to avoid the chance of double free.